torch/csrc/api/include/torch/nn/module.h

```diff
@@ -113,25 +113,39 @@ class CloneableModule : public Module {
  public:
   using Module::Module;
 
-  // should it also detach the gradients, like a deepcopy? Or maybe let's just
-  // give clone() a boolean for this?
-  std::unique_ptr<Module> clone() const override {
-    auto ptr = std::unique_ptr<Module>(
-        new Derived(*static_cast<const Derived*>(this)));
+  virtual void reset() = 0;
+
+  std::shared_ptr<Derived> build() {
+    auto module = std::make_shared<Derived>(static_cast<Derived&&>(*this));
+    module->reset();
+    return std::move(module);
```

Comment:
Return from `std::move` is an anti-pattern (prevents RVO)

Reply:
Also, why does `build()` clone the module instead of modifying it in place? Seems quite wasteful, and might be surprising for people

Reply:
Sorry this was from when I still had it as `unique_ptr`, where compilers are often not smart enough to always elide the return. It should be removed now.

Reply:
If I understand correctly, the answer to the second is that we need to return a shared pointer because all our other code will expect to deal with shared_ptrs

Reply:
I can check whether we can avoid having shared_ptrs here, but e.g. optimizers will want them

Reply:
@apaszke The reason is that the modules could be stack allocated, but to always keep them boxed, they must be moved to the heap at some point.

Reply:
Ohhhh I see. Makes sense. Something that would be great for this API would be to only have to call `.build()` once on the outermost container, and that would make it recursively call it for children. This however would require placing the modules in shared pointers right from the beginning.

Reply:
@apaszke The idiom here is mostly to support this syntax (which is not exactly idiomatic C++, but I see why they want it):

```
std::shared_ptr<RNN> mod = RNN().layers(2).bidirectional(true).make();
```

Idiomatic C++ is to use `std::make_shared` to create a shared pointer to a module, e.g., as in

```
std::shared_ptr<RNN> mod = std::make_shared<RNN>(2, true);
```

but now you don't have a good time to stick on the keyword arguments; if you do it post-facto you need to "initialize" the parameters with an explicit call:

```
// This particular syntax needs enable_shared_from_this
std::shared_ptr<RNN> mod = std::make_shared<RNN>()->layers(2)->bidirectional(true)->init();
```

or fix the implementation of RNN to lazily initialize the parameters upon its first invocation of forward.

I have argued in the past that it's not a big deal to have to write this extra bit of code, but besides being unidiomatic I do not see too many downsides of the idiom advocated here.

Reply:
I only hope that the "expected `shared_ptr` got `Module`" error messages will be readable enough to make people realize that they missed the `.build()`.